### PR TITLE
Implemented user registration

### PR DESCRIPTION
**Pull request by OlexVoit**
Added RegistrationException(in CustomGlobalExceptionHandler)/AuthenticationController(endpoint - registerUser )/User(entity)/UserRepository/UserService(method register)/mapper + dto (for user).

### CODE REVIEW

File: src/main/java/mate/academy/bookshop/dto/user/UserRegistrationRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+@Data
+@FieldMatch
+public class UserRegistrationRequestDto {
+    private static final String NOT_BE_NULL = "Don't be null";
+    private static final String FROM_8_TO_24_CHARACTERS = "Must contain 8 to 24 characters long";
+    private static final String INCORRECT_EMAIL_FORMAT = "Incorrect email format";
+
+    @Email(message = INCORRECT_EMAIL_FORMAT)
+    @NotBlank(message = NOT_BE_NULL)
```
Comment (bhdnchui) [inline src/main/java/mate/academy/bookshop/dto/user/UserRegistrationRequestDto.java:17]: Do we need `@NotBlank` if we have `@Email`?

File: src/main/java/mate/academy/bookshop/exception/CustomGlobalExceptionHandler.java [hunk 2]
```diff
@@ -53,4 +53,13 @@ protected ResponseEntity<Object> handleEntityNotFoundException(Exception ex) {
...
         return new ResponseEntity<>(body, HttpStatus.NOT_FOUND);
     }
+
+    @ExceptionHandler(RegistrationException.class)
+    protected ResponseEntity<Object> handleRegistrationException(Exception ex) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.BAD_REQUEST);
+        body.put("message", ex.getMessage());
+        return new ResponseEntity<>(body, HttpStatus.NOT_FOUND);
```
Comment (bhdnchui) [inline src/main/java/mate/academy/bookshop/exception/CustomGlobalExceptionHandler.java]: Here body is pretty equal to other handlers, please avoid code duplication

File: src/main/java/mate/academy/bookshop/model/Book.java [hunk 3]
```diff
@@ -42,3 +42,6 @@ public class Book {
     @Column(name = "is_deleted", nullable = false)
     private boolean isDeleted = false;
 }
+
+
+
```
Comment (bhdnchui) [inline src/main/java/mate/academy/bookshop/model/Book.java]:
```suggestion
```

File: src/main/java/mate/academy/bookshop/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
+            throw new RegistrationException(CAN_NOT_REGISTER_USER_BY_EMAIL + requestDto.getEmail());
+        }
+        User user = new User();
+        user.setEmail(requestDto.getEmail());
+        user.setPassword(requestDto.getPassword());
+        user.setFirstName(requestDto.getFirstName());
+        user.setLastName(requestDto.getLastName());
+        user.setShippingAddress(requestDto.getShippingAddress());
```
Comment (bhdnchui) [inline src/main/java/mate/academy/bookshop/service/impl/UserServiceImpl.java]: Use mapper for that logic

File: src/main/java/mate/academy/bookshop/validation/PasswordMatchesValidator.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+import mate.academy.bookshop.dto.user.UserRegistrationRequestDto;
+
+public class PasswordMatchesValidator implements ConstraintValidator<FieldMatch,
+        UserRegistrationRequestDto> {
+    @Override
+    public boolean isValid(UserRegistrationRequestDto user,
+                           ConstraintValidatorContext constraintValidatorContext) {
+        return user.getPassword() != null && user.getPassword().equals(user.getRepeatPassword());
```
Comment (bhdnchui) [inline src/main/java/mate/academy/bookshop/validation/PasswordMatchesValidator.java]: Use [Objects#equals](https://www.baeldung.com/java-comparing-objects#equals-static).

**Review by bhdnchui — CHANGES_REQUESTED** [review]
Good job! Just a few points to consider.

**Review by bhdnchui — CHANGES_REQUESTED** [review]
One point to consider.

File: src/main/java/mate/academy/bookshop/exception/CustomGlobalExceptionHandler.java [hunk 6]
```diff
@@ -53,4 +53,13 @@ protected ResponseEntity<Object> handleEntityNotFoundException(Exception ex) {
...
         return new ResponseEntity<>(body, HttpStatus.NOT_FOUND);
     }
+
+    @ExceptionHandler(RegistrationException.class)
+    protected ResponseEntity<Object> handleRegistrationException(Exception ex) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.BAD_REQUEST);
+        body.put("message", ex.getMessage());
+        return new ResponseEntity<>(body, HttpStatus.BAD_REQUEST);
```
Comment (bhdnchui) [inline src/main/java/mate/academy/bookshop/exception/CustomGlobalExceptionHandler.java]: Body is pretty equal to handleEntityNotFoundException method body. Avoid it

File: src/main/java/mate/academy/bookshop/exception/CustomGlobalExceptionHandler.java [hunk 7]
```diff
@@ -47,10 +47,19 @@ private String getErrorMessage(ObjectError error) {
 
     @ExceptionHandler(EntityNotFoundException.class)
     protected ResponseEntity<Object> handleEntityNotFoundException(Exception ex) {
+        return new ResponseEntity<>(getBody(ex, HttpStatus.NOT_FOUND), HttpStatus.NOT_FOUND);
+    }
+
+    @ExceptionHandler(RegistrationException.class)
+    protected ResponseEntity<Object> handleRegistrationException(Exception ex) {
+        return new ResponseEntity<>(getBody(ex, HttpStatus.BAD_REQUEST), HttpStatus.BAD_REQUEST);
```
Comment (yevheniisynytsia) [inline src/main/java/mate/academy/bookshop/exception/CustomGlobalExceptionHandler.java]: writing response code twice seem inconvenient
can we create a method that will get all the required parameters and return a ResponseEntity object?

**Review by yevheniisynytsia — CHANGES_REQUESTED** [review]
good job! let's improve code quality in the exception handler!

File: src/main/java/mate/academy/bookshop/exception/CustomGlobalExceptionHandler.java [hunk 8]
```diff
@@ -45,12 +45,18 @@ private String getErrorMessage(ObjectError error) {
...
+    }
+
+    private Map<String, Object> getBody(Exception ex, HttpStatus status) {
         Map<String, Object> body = new LinkedHashMap<>();
         body.put("timestamp", LocalDateTime.now());
-        body.put("status", HttpStatus.NOT_FOUND);
+        body.put("status", status);
         body.put("message", ex.getMessage());
-        return new ResponseEntity<>(body, HttpStatus.NOT_FOUND);
+        return body;
```
Comment (d1sam) [inline src/main/java/mate/academy/bookshop/exception/CustomGlobalExceptionHandler.java]: writing response code TWICE seem inconvenient. Please create a method that will get all the required parameters and return a ResponseEntity object?

**Review by d1sam — CHANGES_REQUESTED** [review]
Still need to improve code quality in the exception handler!

File: src/main/java/mate/academy/bookshop/exception/CustomGlobalExceptionHandler.java [hunk 9]
```diff
@@ -45,12 +45,15 @@ private String getErrorMessage(ObjectError error) {
         return error.getDefaultMessage();
     }
 
-    @ExceptionHandler(EntityNotFoundException.class)
-    protected ResponseEntity<Object> handleEntityNotFoundException(Exception ex) {
+    @ExceptionHandler({EntityNotFoundException.class, RegistrationException.class})
+    protected ResponseEntity<Object> handleException(Exception ex) {
+        HttpStatus status = (ex instanceof EntityNotFoundException)
+                ? HttpStatus.NOT_FOUND : HttpStatus.BAD_REQUEST;
```
Comment (d1sam) [inline src/main/java/mate/academy/bookshop/exception/CustomGlobalExceptionHandler.java:51]: it's also not the best way, but now we can approve it.

**Review by d1sam — APPROVED** [review]
well done!

**Pull request closed without merging**